### ISSUE TITLE
Sharing an image with the app does not work

### ISSUE THREAD

**Issue by rhrunningrhino** [laurent22/joplin#8870]
Referencing the closed issue #8533 

Hello,
would like to add more info so that the chance to get this bug fixed could be increased:

Device https://github.com/laurent22/joplin/issues/1: Samsung Galaxy S10+
Android: Android 11 running One UI 3.1
Joplin: 2.12.2

Device https://github.com/laurent22/joplin/issues/2: Samsung Galaxy A33
Android: Android 12 running One UI 4
Joplin: 2.12.2

Joplin settings:
Resize large image: "Ask" or "Always" (but doesn't matter)

How to reproduce:

Sharing screen capture images from Samsung Gallery App v14.1.04.2
Share immediately from Screen capture
How to successfully share:

Share from Whatsapp chat, open image share from WhatsApp
Sharing photos taken by Camera, using Samsung Gallery App v14.1.04.2. Resize settings doesn't matter. Resized photo can be successfully shared.
Sharing images on a webpage from Any browser
My Guess:
Originally i thought it is a One UI specific issue but found that there are a few bug reports. The above video is a Xiaomi device, so it should be other issue.

I checked from log, all failed attempts use content:// for filepath, while all successful attempts comes with file://

It appears to me that Joplin can take file:// smoothly but failed to handle content://

Hope that helps! If this bug can be fixed, I will definitely leaving Evernote!!

**Comment by wh201906** [comment]
@jd1378 Hi. Would you like to take a look at this?

**Comment by wh201906** [comment]
I did some tests and found this bug might be introduced in v2.10.5
I tested v2.10.1, v2.10.4 and they work. But in v2.10.5 I got this error

<details><summary>Log</summary>

```
09-24T23:08:19,20,"""Could not attach file:", "Error: no such column: document_id (code 1): , while compiling: SELECT document_id, _display_name, mime_type, _size, last_modified FROM image WHERE _id=?
Code: EUNSPECIFIED
Error: no such column: document_id (code 1): , while compiling: SELECT document_id, _display_name, mime_type, _size, last_modified FROM image WHERE _id=?
    at promiseMethodWrapper (address at index.android.bundle:1:111187)
    at copyFile (address at index.android.bundle:1:2939183)
    at ?anon_0_ (address at index.android.bundle:1:2936520)
    at next (native)
    at fulfilled (address at index.android.bundle:1:2932494)
    at tryCallOne (/root/react-native/ReactAndroid/hermes-engine/.cxx/MinSizeRel/164673tg/arm64-v8a/lib/InternalBytecode/InternalBytecode.js:53:16)
    at anonymous (/root/react-native/ReactAndroid/hermes-engine/.cxx/MinSizeRel/164673tg/arm64-v8a/lib/InternalBytecode/InternalBytecode.js:139:27)
    at apply (native)
    at anonymous (address at index.android.bundle:1:213202)
    at _callTimer (address at index.android.bundle:1:212151)
    at _callReactNativeMicrotasksPass (address at index.android.bundle:1:212315)
    at callReactNativeMicrotasks (address at index.android.bundle:1:214290)
    at __callReactNativeMicrotasks (address at index.android.bundle:1:114674)
    at anonymous (address at index.android.bundle:1:113767)
    at __guard (address at index.android.bundle:1:114515)
    at flushedQueue (address at index.android.bundle:1:113678)
    at invokeCallbackAndReturnFlushedQueue (address at index.android.bundle:1:113621)""
09-24T23:08:19,30,"""Original dimensions ", "{"width":123,"height":123}""
09-24T23:08:18,30,"""DecryptionWorker: cannot start because no master key is currently loaded.""
09-24T23:08:18,30,"""Got file: content://com.miui.gallery.open/raw/%2Fstorage%2F6666-6666%2FDCIM%2FCamera%2Ftesttesttest.jpg""
09-24T23:08:18,30,"""Got type: image/jpeg""
09-24T23:08:18,30,"""about to attach resource {"name":"testtesttest.jpg","mimeType":"image/jpeg","uri":"content://com.miui.gallery.open/raw/%2Fstorage%2F6666-6666%2FDCIM%2FCamera%2Ftesttesttest.jpg"}""
09-24T23:08:18,30,"""Received shared data""
```

</details> 

I guess this bug is introduced in #7637  
v2.10.6 reports the same error (no such column), but in v2.10.7 and v2.12.1, it reports `file does not exist at...`, just like the error in #8533. I have no idea why the behavior changed but I guess the error `file does not exist at...` is caused by the failed query.

**Comment by jd1378** [comment]
hi
these are the two PRs that were related to share: #7799 and #7807 
and as far as I remember, they were working correctly at the time of being merged

about the log, I believe I had applied a fix that was handling missing columns in #7801 , if this log is related to my module
though a missing document id is not what I expected to happen at least, so that may need small amount of work

edit: removed the dumbness lol

**Comment by wh201906** [comment]
By the way, all of my file explorers(the built-in one of MIUI and [MaterialFiles](https://github.com/zhanghai/MaterialFiles)) can share images to Joplin. The URIs of them also start with `content://`

**Comment by wh201906** [comment]
> weird thing is I can't see the changes from #7801 [on the dev branch](https://github.com/jd1378/joplin/blob/dev/packages/react-native-saf-x/android/src/main/java/com/reactnativesafx/utils/DocumentStat.java) (?!)

I can see the changes from #7801
https://github.com/laurent22/joplin/blob/fb9974c76ffd94a50deb56783314198f8d74c781/packages/react-native-saf-x/android/src/main/java/com/reactnativesafx/utils/DocumentStat.java#L58-L74

The `dev branch` you mentioned is in your own repo(jd1378/joplin/tree/dev), not in this one(laurent22/joplin/tree/dev). I guess you are writing code on a new branch of your own repo(jd1378/joplin/tree/fix_safx_handle_missing_data) for #7801

**Comment by jd1378** [comment]
ooooh my bad, don't know how I missed that
thanks for pointing it out, but I feel dumb now 😅

can you check the log prior to the error and tell me which content uri is exactly causing the error ?

**Comment by jd1378** [comment]
I can successfully share screenshots immediately to joplin on miui 14

whats the steps to reproduce on your phone ?

**Comment by wh201906** [comment]
Yesterday I checked the implementation of `DocumentHelper.java` and `EfficientDocumentHelper.java` in #7637. I guess the [`getDocumentUri()`](https://github.com/laurent22/joplin/blob/b82bf165059e6737768430902b3514dd28688d9a/packages/react-native-saf-x/android/src/main/java/com/reactnativesafx/utils/EfficientDocumentHelper.java#L69) is the replacement of [`goToDocument()`](https://github.com/laurent22/joplin/blob/2f254d81cd2380afd5b3f1c6a5473670435fcb62/packages/react-native-saf-x/android/src/main/java/com/reactnativesafx/utils/DocumentHelper.java#L432), in which you use [`getStat()`](https://github.com/laurent22/joplin/blob/b82bf165059e6737768430902b3514dd28688d9a/packages/react-native-saf-x/android/src/main/java/com/reactnativesafx/utils/EfficientDocumentHelper.java#L229) to check if the file exists.

https://github.com/laurent22/joplin/blob/b82bf165059e6737768430902b3514dd28688d9a/packages/react-native-saf-x/android/src/main/java/com/reactnativesafx/utils/EfficientDocumentHelper.java#L103
https://github.com/laurent22/joplin/blob/b82bf165059e6737768430902b3514dd28688d9a/packages/react-native-saf-x/android/src/main/java/com/reactnativesafx/utils/EfficientDocumentHelper.java#L147-L152

However, in `DocumentHelper.java`, you don't check the existence when `pathSegmentsToTraverseLength == 0`. And the check for uri with `content://` actually doesn't work.
https://github.com/laurent22/joplin/blob/2f254d81cd2380afd5b3f1c6a5473670435fcb62/packages/react-native-saf-x/android/src/main/java/com/reactnativesafx/utils/DocumentHelper.java#L464-L472 
The `fromSingleUri()` returns null only when the SDK version is lower than 19, and the constructor of `SingleDocumentFile` doesn't check the existence.
https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:documentfile/documentfile/src/main/java/androidx/documentfile/provider/DocumentFile.java;l=116
https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:documentfile/documentfile/src/main/java/androidx/documentfile/provider/SingleDocumentFile.java;l=32
https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:documentfile/documentfile/src/main/java/androidx/documentfile/provider/DocumentFile.java;l=88

In short, the code before PR #7801 doesn't check the existence of a file from uri, that's why v2.10.4 and older versions work.

**Comment by wh201906** [comment]
> can you check the log prior to the error and tell me which content uri is exactly causing the error ?

It's in the log from https://github.com/laurent22/joplin/issues/8870#issuecomment-1732597883

`content://com.miui.gallery.open/raw/%2Fstorage%2F6666-6666%2FDCIM%2FCamera%2Ftesttesttest.jpg`

**Comment by wh201906** [comment]
> I can successfully share screenshots immediately to joplin on miui 14
> 
> whats the steps to reproduce on your phone ?

Yes I can also share a screenshot to Joplin immediately. But if I open the screenshot later in the MIUI Gallery then send it, the bug occurs.
The uri of former starts with `content://com.miui.screenshot.provider.fileProvider/raw/`, and the uri of latter starts with `content://com.miui.gallery.open/raw/`. They are from different apps so I guess that's why the behaviors differ.

**Comment by jd1378** [comment]
I see
thanks a lot for the inspection, that helps
I'm sorry but I'm very busy at the moment and I'm not sure if I can tend to this any time soon (not for a week or two at least)

**Comment by wh201906** [comment]
@jd1378 Would you mind if I make a PR to fix it? I have made a prototype and it does fix this issue.

**Comment by jd1378** [comment]
@wh201906 not at all, I'm actually happy that you worked on it already and made a PR, as I am very busy (I appreciate it even if I wasn't busy)
Thank you
